### PR TITLE
Fix issue 9029: make all types (including basic) match to alias template parameters with priority MATCH.convert.

### PR DESCRIPTION
**Pull request by FeepingCreature**
This may break code. The point of this PR is partially to determine if it does, and if so what.

This is not done by inserting basic types into the symbol table, but simply passing them through in the compiler as objects. This is the main reason I am unconfident about this PR.

### CODE REVIEW

**Comment by FeepingCreature** [comment]
Hm, this seems to need some fixes. Should it go behind `-experimental or should I do a full DIP writeup?

**Comment by FeepingCreature** [comment]
See https://github.com/dlang/phobos/compare/master...FeepingCreature:fix/issue-9029-fixes for required Phobos changes.

File: test/compilable/ddocunittest.d
Comment (jacob-carlborg) [inline test/compilable/ddocunittest.d]: Phobos should ideally not be used in the tests.

Comment (FeepingCreature) [reply test/compilable/ddocunittest.d]: Oh oops. Hm.

**Comment by jacob-carlborg** [comment]
No other tests?

**Comment by FeepingCreature** [comment]
I should add one, shouldn't I. Hang on.

File: test/compilable/ddocunittest.d
Comment (FeepingCreature) [reply test/compilable/ddocunittest.d]: Fixed?

**Comment by FeepingCreature** [comment]
I put it behind `-preview=allowTypesForTemplateAlias`.

**Comment by jacob-carlborg** [comment]
Why is the flag necessary?

**Comment by FeepingCreature** [comment]
Because it potentially "breaks" existing code due to templates matching that didn't used to match, making template instantiation ambiguous. You can see instances of that in the Phobos PR.

**Review by jacob-carlborg — APPROVED** [review]

**Comment by jacob-carlborg** [comment]
It might be worth having an explicit changelog entry for this enhancement.

**Comment by TurkeyMan** [comment]
Ethan will be so happy!

File: src/dmd/dtemplate.d
Comment (Geod24) [inline src/dmd/dtemplate.d]: `sa` is an `extern(C++)` class, hence the third condition (`cast(Type) sa`) does not do a dynamic cast, but a reinterpret one.

Comment (FeepingCreature) [reply src/dmd/dtemplate.d]: Oh f*ck.

Comment (FeepingCreature) [reply src/dmd/dtemplate.d]: Better?

**Comment by FeepingCreature** [comment]
Added a changelog entry, fixed a rather embarrassing bug meaning that the alias check matched on every possible parameter, which reminds me that I can't think of any type, value or other thing that alias *doesn't* match now.

**Comment by wilzbach** [comment]
Now that the Phobos PR is merged, how about temporarily enabling this by default and checking the output of the project tester?
Because maybe we don't need a preview flag at all.

**Comment by FeepingCreature** [comment]
Optimistic! Let's try it!

**Comment by FeepingCreature** [comment]
_Damn_. That's _so_ close.

**Comment by JinShil** [comment]
> Can anyone poke the Weka guys about looking at this?

Create an issue in their repository.

File: changelog/template_alias_matches_basic_types.dd
Comment (JohanEngelen) [inline changelog/template_alias_matches_basic_types.dd]: Extra points for showing what cases break, and _how to fix it_. :)

Comment (FeepingCreature) [reply changelog/template_alias_matches_basic_types.dd]: Better?

Comment (JohanEngelen) [reply changelog/template_alias_matches_basic_types.dd]: thanks. Perhaps add explicitly that the code would break for delegates.
I guess things can break in many different ways. Regardless, the example shows the type of fix that is required.

**Comment by FeepingCreature** [comment]
Now that I've fixed issues in Phobos as well as Mecca relating to this, some thoughts.

There was a happy coincidence with alias in that "the sort of things that could be passed to alias" were exactly "the sort of things you might want to, _in some abstract sense_, call." in other words, lambdas, nested functions, structs, and strings, but specifically _not_ things like delegate types. This allowed for easy overloading between `alias A` and `T`, where `T` would represent a runtime callable parameter and `A` a compiletime callable parameter. However, this was of course coincidental.

Now that `alias` can match every type, this is awkward. There is no check in Phobos `std.traits` that encapsulates what `alias` used to do. One might reasonably expect `isCallable` to do the job, but for unclear reasons it is `true` for `void delegate()`, which is _not_ callable in the "can put () on it" sense. And of course, structs complicate this further by being either compiletime or runtime callable.

Is `__traits(compiles, A(...))` sufficient? I'm not sure.

**Comment by jacob-carlborg** [comment]
> edit: facepalm it's using the tagged version of mecca...

Yeah, they need to create a new tag.

**Comment by jacob-carlborg** [comment]
When this is merge, please make sure the commits are squashed.

**Comment by jacob-carlborg** [comment]
> One might reasonably expect isCallable to do the job, but for unclear reasons it is true for void delegate(), which is not callable in the "can put () on it" sense

Why is `void delegate()` not callable?

**Comment by wilzbach** [comment]
@FeepingCreature you are already a member of the Buildkite Organization. This means you can login there and then just click on rebuild for the individual test that failed. Saves you a few commits:

image

Also, as these CI questions come up often in Slack, let me know if you want to have an invite.

**Comment by FeepingCreature** [comment]
> When this is merge, please make sure the commits are squashed.

Noted and aware. That's why it has that marker on the title.

> @FeepingCreature you are already a member of the Buildkite Organization.

Augh I keep forgetting. _slaps head_

> Why is void delegate() not callable?

_A delegate_ is callable. _`void delegate()`_, on the other hand, is not callable. `alias A = void delegate(); static assert(is(A()));`

**Comment by wilzbach** [comment]
> why it has that marker on the title.

yYou might want to give draft PRs a try next time: https://github.blog/2019-02-14-introducing-draft-pull-requests/

**Comment by JohanEngelen** [comment]
@FeepingCreature I very much appreciate the thoroughness. It is indeed a pity that there is currently no existing trait that would restore prior `alias` behaviour.
I don't understand how you would call `"a > b"` by just putting `()` after it?

**Comment by FeepingCreature** [comment]
Right, you don't; it's more correct to say that strings like "a > b" happen to be the sort of thing that "would be passed to `std.functional.unaryFun`/`std.functional.binaryFun` so maybe the correct approach is checking for those explicitly.

**Comment by jacob-carlborg** [comment]
> Noted and aware. That's why it has that marker on the title.

It would be more clear if you created fixup commits instead. There are quite a few contributors that don't know that commits should be amended or squashed. I can't keep track of which contributors know this and which don't, therefore I'm pointing it out as I see it.

**Comment by FeepingCreature** [comment]
I'm only holding back on that because I don't know whether we want the flag or not yet, so I don't want to destroy the commit that had the flag until it's time to push for merging. But thanks for the info.

**Comment by wilzbach** [comment]
If it passes on all CIs including the project tester, I don't think we need the special flag and can make Ethan immediately happy with the next release.

**Comment by GooberMan** [comment]
Yee. And just chatting on IRC about it, it might be best to do a blog post or something with the release and explain that your code that fails to compile can be fixed by deleting lines/inserting template constraints/etc.

**Comment by wilzbach** [comment]
> it might be best to do a blog post or something with the release

FYI @FeepingCreature et.al. Mike Parker compiles the release blog post based on changelog entries he sees. So just make sure that the changelog entry is long and detail enough, s.t. it does the following:

> explain that your code that fails to compile can be fixed by deleting lines/inserting template constraints/etc.

Though as far as I can tell the changelog entry looks pretty complete and detailed to me. @GooberMan anything that you miss in there as you brought this up?

**Comment by FeepingCreature** [comment]
Flag removed. Ready on my end.

**Comment by GooberMan** [comment]
@wilzbach - Chanelog looks good to me. Although I still expect a bit of initial confusion.

File: test/compilable/ddocunittest.d
Comment (thewilsonator) [inline test/compilable/ddocunittest.d]: @FeepingCreature Still need to remove this use of flag.

File: test/compilable/test9029.d
Comment (thewilsonator) [inline test/compilable/test9029.d]: And this one.

**Review by thewilsonator — APPROVED** [review]
Otherwise LGTM.

File: test/compilable/ddocunittest.d
Comment (FeepingCreature) [reply test/compilable/ddocunittest.d]: fixed

File: test/compilable/test9029.d
Comment (FeepingCreature) [reply test/compilable/test9029.d]: fixed thanks

File: test/compilable/test9029.d
Comment (wilzbach) [inline test/compilable/test9029.d]: All test files should have a `TEST_OUTPUT` section (even if empty) to avoid that deprecations or warnings are silently ignored.

Comment (FeepingCreature) [reply test/compilable/test9029.d]: fixed

**Comment by GooberMan** [comment]
https://www.youtube.com/watch?v=z_5J1csi1AU

(The Dance of Joy for those that don't follow links)

**Comment by TurkeyMan** [comment]
Wow... that just happened.
Who wants to do a PR to remove read/write access from `shared`? ;)

**Comment by FeepingCreature** [comment]
I wish somebody would make a PR to remove `shared` and salt the earth...

**Comment by TurkeyMan** [comment]
O_o

**Comment by quickfur** [comment]
This PR has caused a regression: https://issues.dlang.org/show_bug.cgi?id=19884

**Pull request merged**
